File: app/src/main/java/com/example/my_application114514/service/MusicService.java
package com.example.my_application114514.service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import android.app.Service;
import android.content.Intent;
import android.content.res.AssetFileDescriptor;
import android.content.res.AssetManager;
import android.media.MediaPlayer;
import android.os.Binder;
import android.os.IBinder;

import com.example.my_application114514.data.SongData;
import com.example.my_application114514.listener.PlayerListener;
import com.example.my_application114514.util.MusicBinder;
import com.example.my_application114514.util.PlayModeHelper;

import androidx.annotation.Nullable;

public class MusicService extends Service {

  private MediaPlayer mMediaPlayer;
  public ArrayList<SongData> mPlaylist;
  public int curIndex;

  int mCurMode = PlayModeHelper.PLAY_ORDERED;  // 播放模式
  private int randomCnt = -1;  // 随机计数器
  private int[] randomOrder;  // 随机播放顺序数组

  private PlayerListener mPlayerListener;

  @Override
  public void onCreate() {
    super.onCreate();
    mMediaPlayer = new MediaPlayer();
    mMediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
      @Override
      public void onCompletion(MediaPlayer mp) {
        switch (mCurMode) {
          case PlayModeHelper.PLAY_ORDERED:
            orderedMode();break;
          case PlayModeHelper.PLAY_CIRCLE:
            circleMode();break;
          case PlayModeHelper.PLAY_SINGLE:
            singleMode();break;
          case PlayModeHelper.PLAY_RANDOM:
            randomMode();break;
          default:break;
        }

        // 自然播放结束到下一首后，通知主活动更新界面
        if(mPlayerListener != null){
          mPlayerListener.onComplete(curIndex,mPlaylist.get(curIndex));
        }

      }
    });

    mPlaylist = new ArrayList<>();
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    if(mMediaPlayer!=null){
      mMediaPlayer.stop();
      mMediaPlayer.release();
      mMediaPlayer = null;
    }
  }

  @Override
  public boolean onUnbind(Intent intent) {
    return super.onUnbind(intent);
  }

  @Override
  public int onStartCommand(Intent intent, int flags, int startId) {
    return super.onStartCommand(intent, flags, startId);
  }

  @Nullable
  @Override
  public IBinder onBind(Intent intent) {
    return new MusicBinder(this); // 返回binder的同时，通过binder的构造函数与service绑定
  }

  public void updateMusicList(ArrayList<SongData> mList){
    mPlaylist = mList;
  }

  public void updateMusicCurIndex(int idx){curIndex = idx;}

  public void startPlay(){
    if(curIndex<0 || curIndex >= mPlaylist.size()){
      return;
    }

    SongData song = mPlaylist.get(curIndex);
    String songName = song.getSongName();

    AssetManager assetManager = getAssets();

    try{
      mMediaPlayer.stop();
      mMediaPlayer.reset();
      AssetFileDescriptor fileDescriptor = assetManager.openFd(songName);
      mMediaPlayer.setDataSource(fileDescriptor.getFileDescriptor(),
          fileDescriptor.getStartOffset(),
          fileDescriptor.getLength());
      mMediaPlayer.prepare();
      mMediaPlayer.start();

    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public boolean isPlaying() {
    return mMediaPlayer.isPlaying();
  }

  public void pause() {
    if(!mMediaPlayer.isPlaying()){
      return;
    }
    mMediaPlayer.pause();
  }

  public void play() {
    if(mMediaPlayer.isPlaying()){
      return;
    }
    mMediaPlayer.start();
  }

  public void last(){
    int preIdx = (curIndex-1 + mPlaylist.size()) % mPlaylist.size();
    playIdx(preIdx);
  }

  public void next(){
    int nxtIdx = (curIndex+1) % mPlaylist.size();
    playIdx(nxtIdx);
  }

  public void playIdx(int position){
    updateMusicCurIndex(position);
    startPlay();
  }

  public void stop() {
    mMediaPlayer.stop();
    try{
    mMediaPlayer.prepare();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  public int getCurProgress() {
    return mMediaPlayer.getCurrentPosition();
  }

  public int getEndProgress(){
    return mMediaPlayer.getDuration();
  }

  public void seekTo(int progress) {
    mMediaPlayer.seekTo(progress);
  }

  public void setPlayMode(int mode) {
    mCurMode = mode;
  }

  public void setPlayerListener(PlayerListener playerListener){
    this.mPlayerListener = playerListener;
  }

  // ====================定义播放模式方法====================

  void orderedMode(){ if(curIndex+1 != mPlaylist.size()){next();} }
  void circleMode(){ next(); }
  void singleMode(){ playIdx(curIndex); }
  void randomMode(){
    // 更新随机播放列表
    if(randomCnt == -1 || randomCnt == mPlaylist.size()){
      randomCnt = 0;
      randomOrder = getRandomPermutation(mPlaylist.size());
    }

    int nextIdx = randomOrder[randomCnt++];
    playIdx(nextIdx);
  }


  public static int[] getRandomPermutation(int n) {
    // 创建一个包含 0 到 n-1 的列表
    List<Integer> list = IntStream.range(0, n).boxed().collect(Collectors.toList());
    // 使用 Collections.shuffle 方法对列表进行随机排序
    Collections.shuffle(list);
    // 将列表转换为 int 数组
    return list.stream().mapToInt(Integer::intValue).toArray();
  }




}

File: app/src/main/java/com/example/my_application114514/service/MusicService.kt
package com.example.my_application114514.service

import android.app.Service
import android.content.Intent
import android.content.res.AssetFileDescriptor
import android.content.res.AssetManager
import android.media.MediaPlayer
import android.os.IBinder
import android.util.Log
import com.example.my_application114514.R
import com.example.my_application114514.data.GlobalConsts
import com.example.my_application114514.data.SongData
import com.example.my_application114514.listener.MyPlayerListener
import com.example.my_application114514.service.binder.MusicBinder
import com.example.my_application114514.util.InteractiveEvent
import java.io.IOException

class MusicService : Service() {

    private lateinit var mMyPlayerListener : MyPlayerListener
    private lateinit var mMediaPlayer: MediaPlayer
    private var mRamdomPlayList: ArrayList<Int> ?= ArrayList()
    private var mSongList : ArrayList<SongData> ?= ArrayList()
    private var curSongIdx : Int = 0
    private var mPlayMode : Int = GlobalConsts.PLAY_MODE_ORDERED
    private var mRndIdx : Int = 0


    override fun onCreate() {
        super.onCreate()
        mMediaPlayer = MediaPlayer()
        mSongList = ArrayList()

        // 添加播放完成监听器
        mMediaPlayer.setOnCompletionListener {
            onMusicCompletion()
        }
    }

    override fun onBind(intent: Intent?): IBinder? {
        return MusicBinder(this)
    }

    // 新增：播放完成回调处理
    private fun onMusicCompletion(){
        if (mSongList.isNullOrEmpty()) return
        when(mPlayMode){
            GlobalConsts.PLAY_MODE_CIRCLE  -> { nextSong() }
            GlobalConsts.PLAY_MODE_RANDOM  -> { randomPlay() }
            GlobalConsts.PLAY_MODE_SINGLE  -> { startPlay(curSongIdx,true) }
            GlobalConsts.PLAY_MODE_ORDERED -> {
                if(curSongIdx + 1 != mSongList?.size)nextSong()
                else{
                    mMediaPlayer.stop()
                    mMyPlayerListener?.onStop()
                }
            }
            else -> { nextSong() }
        }
    }

    // ============================== 与binder交互的部分
    fun setMusicList(data: ArrayList<SongData>?) { this.mSongList = data }
    fun setMusicCurIndex(idx: Int) { this.curSongIdx = idx }
    fun setPlayerListener(myPlayerListener: MyPlayerListener){this.mMyPlayerListener = myPlayerListener}
    fun getMusicList():ArrayList<SongData>? { return this.mSongList  }
    fun getMusicCurIndex():Int { return this.curSongIdx }
    fun getCurProgress():Int{return mMediaPlayer.currentPosition}
    fun getPlayMode():Int{return mPlayMode}


    // ==================== 播放模式控制
    fun switchMode(){
        mPlayMode = (mPlayMode+1) % 4
    }

    // 生成随机模式播放列表
    fun setRandonList(){
        mRamdomPlayList?.clear()
        mRamdomPlayList = InteractiveEvent.setRandomList(mSongList)
    }

    // 传入随机序号
    fun randomPlay(){
        if(mRamdomPlayList?.size == 0) { setRandonList() }
        val inputIdx = mRamdomPlayList?.get(mRndIdx) as Int
        startPlay(inputIdx)

        mRndIdx += 1
        if(mRndIdx == mSongList?.size){
            setRandonList()
            mRndIdx = 0
        }
    }


    // ==================== 播放控制
    fun seekTo(progress:Int){ mMediaPlayer.seekTo(progress) }
    fun isMusicPlaying() : Boolean{ return mMediaPlayer.isPlaying() }
    fun pause(){ mMediaPlayer.pause() }
    fun stop(){
        mMediaPlayer.stop()
        mMediaPlayer.prepare()
    }
    fun play(){
        if(mMediaPlayer.isPlaying())return
        mMediaPlayer.start()
    }
    fun previous(){
        var idx = (curSongIdx + (mSongList?.size ?: 1) - 1) % (mSongList?.size ?: 1)
        if(mPlayMode == GlobalConsts.PLAY_MODE_RANDOM) { randomPlay() }
        else { startPlay(idx) }
        mMyPlayerListener?.onPlay()
    }
    fun nextSong(){
        var idx = (curSongIdx + 1) % (mSongList?.size ?: 1)
        if(mPlayMode == GlobalConsts.PLAY_MODE_RANDOM) { randomPlay() }
        else { startPlay(idx) }
        mMyPlayerListener?.onPlay()
    }

    // 播放默认曲目
    fun startPlay(){
        if(mMediaPlayer.isPlaying)return
        val song : SongData = mSongList?.getOrNull(curSongIdx) ?: return
        try {
            val assetManager: AssetManager = this.assets
            val afd: AssetFileDescriptor = assetManager.openFd(song.assetPath)
            mMediaPlayer.setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
            afd.close() // 及时关闭描述符

            mMediaPlayer.prepare()
            mMediaPlayer.start()
            mMediaPlayer.setOnErrorListener { _, what, extra ->
                Log.e("Player", "播放错误: $what | $extra")
                false
            }
        }catch (e: IOException) {
            Log.e("Player", "播放失败", e)
        } catch (e: IllegalArgumentException) {
            Log.e("Player", "无效参数", e)
        }
    }

    // 播放指定曲目
    fun startPlay(idx:Int,forceStopFlag: Boolean = false){
        val needRestart = (idx != curSongIdx) || forceStopFlag  // 判断是否需要暂停播放
        if (needRestart && mMediaPlayer.isPlaying) { mMediaPlayer.stop() }
        if (needRestart) {
            try {
                mMediaPlayer.reset()  // 重置播放器状态
                curSongIdx = idx      // 更新当前曲目索引
            } catch (e: IllegalStateException) {
                Log.e("Player", "重置播放器失败", e)
                return
            }
        }
        startPlay() // 调用无参数版本开始播放
    }

    // 监听并通知


    // 新增：在服务销毁时释放资源
    override fun onDestroy() {
        releaseMediaPlayer()
        super.onDestroy()
    }

    private fun releaseMediaPlayer() {
        if (::mMediaPlayer.isInitialized) {
            try {
                mMediaPlayer.reset()       // 清除播放器状态
                mMediaPlayer.release()     // 释放系统资源
            } catch (e: Exception) {
                Log.e("MusicService", "释放MediaPlayer失败", e)
            }
        }
    }
}




